crates/control/src/motion/walking_engine.rs

```diff
@@ -165,13 +165,24 @@ impl WalkingEngine {
             &context.sensor_data.inertial_measurement_unit,
         );
 
-        if self.t.is_zero() {
-            self.initialize_step_states_from_request(
-                *context.walk_command,
-                context.support_foot.support_side,
-                context.config,
-                context.kick_steps,
-            );
+        let is_step_started_this_cycle = self.t.is_zero();
+        if is_step_started_this_cycle {
+            match context.has_ground_contact {
+                true => {
+                    self.initialize_step_states_from_request(
+                        *context.walk_command,
+                        self.swing_side,
+                        context.config,
+                        context.kick_steps,
+                    );
+                }
+                false => {
+                    self.current_step = Step::zero();
+                    self.step_duration = Duration::ZERO;
+                    self.swing_side = Side::Left;
+                    self.max_swing_foot_lift = 0.0;
+                }
+            }
```

Comment:
Is there a reason why you chose a match statement here over an if?

Reply:
I just found it more readable, because you can differentiate better between the cases of has_ground_contact.

Reply:
I changed it to if else.